storage/grpc_client.go

```diff
@@ -956,3 +1021,303 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
 	r.cancel = cancel
 	return res.response, nil
 }
+
+func newGRPCWriter(c *grpcStorageClient, params *openWriterParams, r io.Reader) *gRPCWriter {
+	size := params.chunkSize
+	if params.chunkSize == 0 {
+		// TODO: Should we actually use the minimum of 256 KB here when the user
+		// indicates they want minimal memory usage? We cannot do a zero-copy,
+		// bufferless upload like HTTP/JSON can.
+		// TODO: We need to determine if we can avoid starting a
+		// resumable upload when the user *plans* to send more than bufSize but
+		// with a bufferless upload.
+		size = maxPerMessageWriteSize
+	}
+
+	return &gRPCWriter{
+		buf:           make([]byte, size),
+		c:             c,
+		ctx:           params.ctx,
+		reader:        r,
+		bucket:        params.bucket,
+		object:        params.object,
+		attrs:         params.attrs,
+		conds:         params.conds,
+		encryptionKey: params.encryptionKey,
+		sendCRC32C:    params.sendCRC32C,
+	}
+}
+
+type gRPCWriter struct {
+	c      *grpcStorageClient
+	buf    []byte
+	reader io.Reader
+
+	ctx context.Context
+
+	bucket        string
+	object        string
+	attrs         *ObjectAttrs
+	conds         *Conditions
+	encryptionKey []byte
+
+	sendCRC32C bool
+
+	// The gRPC client-stream used for sending buffers.
+	//
+	// This is an experimental API and not intended for public use.
```

Comment:
nit: maybe move this comment above the entire type

Reply:
I instead wrote a specific comment because I wanted to keep this one on the property itself.